### PR TITLE
Oauth2 flow support

### PR DESCRIPTION
**Pull request by boajer**
https://oat-sa.atlassian.net/browse/NEX-355
WIP
composer.json should be updated by inclusion of  `"oat-sa/extension-tao-oauth": "dev-develop"`

depends on https://github.com/oat-sa/extension-tao-oauth/pull/39

### CODE REVIEW

File: model/DeliverTenant.php [hunk 1]
```diff
@@ -84,9 +89,30 @@ public function getLtiCredentials()
         return $this->ltiProviders;
     }
 
+    /**
+     * @return OauthCredentials[]|[]
+     * @throws \common_exception_ValidationFailed
```
[the remark this answers is not in the document]
Comment (boajer) [reply model/DeliverTenant.php]: https://github.com/oat-sa/tao-core/blob/fc16939370a463af1a58da6b4f1915b9acf6d529/models/classes/auth/AbstractCredentials.php#L60

File: scripts/install/RegisterServices.php [hunk 2]
```diff
@@ -17,7 +19,17 @@ class RegisterServices extends InstallAction
...
     public function __invoke($params)
     {
-        $publicationService = new PublicationService([PublicationService::OPTION_LOOKUP_INTERVAL => 10, PublicationService::OPTION_MAX_LOOKUP_ATTEMPTS => 100]);
+        $publicationService = new PublicationService(
+            [
+                PublicationService::OPTION_LOOKUP_INTERVAL => 10,
+                PublicationService::OPTION_MAX_LOOKUP_ATTEMPTS => 100,
+                PublicationService::OPTION_API_CLIENT => new RestApiClient([RestApiClient::OPTION_HTTP_CLIENT => new OAuthClient(
+                    [
+                        OAuthClient::OPTION_TOKEN_STORAGE => 'cache',
```
Comment (jbout) [inline scripts/install/RegisterServices.php]: cache might not be the best choice for default persistence, have you considered "default_kv" ?

Comment (mike-ionut-mihai-sandu-tao) [reply scripts/install/RegisterServices.php]: fixed

File: model/DeliverTenant.php [hunk 3]
```diff
@@ -84,9 +89,30 @@ public function getLtiCredentials()
...
-        return $this->hasOption(self::OPTION_OAUTH2_CREDENTIALS) ? $this->getOption(self::OPTION_OAUTH2_CREDENTIALS) : [];
+        if (null === $this->oauthCredentials) {
+
+            if ($this->hasOption(self::OPTION_OAUTH2_CREDENTIALS)) {
+                foreach ($this->getOption(self::OPTION_OAUTH2_CREDENTIALS) as $credentials) {
+                    $this->oauthCredentials[] = new OauthCredentials([
+                        Provider::CLIENT_ID => $credentials['key'],
+                        Provider::CLIENT_SECRET => $credentials['secret'],
+                        Provider::GRANT_TYPE => 'client_credentials',
+                        Provider::TOKEN_URL => $this->getRootUrl() . 'auth/access-tokens',
```
[the remark this answers is not in the document]
Comment (mike-ionut-mihai-sandu-tao) [reply model/DeliverTenant.php]: done

File: model/DeliverTenant.php [hunk 4]
```diff
@@ -84,9 +89,30 @@ public function getLtiCredentials()
...
     }
 
+    /**
+     * @return OauthCredentials[]|[]
+     * @throws \common_exception_ValidationFailed
+     */
     public function getOauth2Credentials()
     {
-        return $this->hasOption(self::OPTION_OAUTH2_CREDENTIALS) ? $this->getOption(self::OPTION_OAUTH2_CREDENTIALS) : [];
+        $rootUrl = $this->getRootUrl() . 'auth/access-tokens';
```
Comment (jbout) [inline model/DeliverTenant.php]: not a root url, either retrieve this value in foreach loop (as it is just reading a value from an array) or call it tokenUrl or accessTokenUrl

File: manifest.php [hunk 5]
```diff
@@ -28,12 +28,13 @@
     'label' => 'Deliver Connect',
     'description' => 'Extension to manage the connection between Tao Backoffice and Nextgen Deliver',
     'license' => 'GPL-2.0',
-    'version' => '0.5.3',
+    'version' => '0.6.0',
     'author' => 'Open Assessment Technologies SA',
     'requires' => array(
         'tao' => '>=38.1.1',
```
Comment (jbout) [inline manifest.php]: please add dependancy to default_kv introduction

File: model/DeliverTenant.php [hunk 6]
```diff
@@ -84,9 +89,30 @@ public function getLtiCredentials()
...
+                foreach ($this->getOption(self::OPTION_OAUTH2_CREDENTIALS) as $credentials) {
+                    $this->oauthCredentials[] = new OauthCredentials([
+                        Provider::CLIENT_ID => $credentials['key'],
+                        Provider::CLIENT_SECRET => $credentials['secret'],
+                        Provider::GRANT_TYPE => 'client_credentials',
+                        Provider::TOKEN_URL => $rootUrl,
+                    ]);
+                }
+            } else {
+                $this->oauthCredentials = [];
```
Comment (jbout) [inline model/DeliverTenant.php]: move to line 101?

File: model/PublicationService.php [hunk 7]
```diff
@@ -74,6 +76,7 @@ public function publish(DeliverTenant $deliverTenant, QtiPackage $qtiPackage)
      * @return LookupSnapshot
      * @throws GuzzleException
      * @throws InconsistentDataException
+     * @throws InvalidServiceManagerException
```
Comment (jbout) [inline model/PublicationService.php]: should not throw InvalidServiceManagerException

File: manifest.php (hunk 5 above)
Comment (mike-ionut-mihai-sandu-tao) [reply manifest.php]: done

File: model/DeliverTenant.php (hunk 6 above)
Comment (mike-ionut-mihai-sandu-tao) [reply model/DeliverTenant.php]: done

File: model/PublicationService.php (hunk 7 above)
Comment (jbout) [reply model/PublicationService.php]: correction, this exception can be thrown so docblock is correct

**Pull request merged**